Add searching to the teams view

andy-armstrong opened the pull request:
https://openedx.atlassian.net/browse/TNL-1935

### Sandbox
- [x] http://andy-armstrong.sandbox.edx.org/courses/course-v1:AndyA+AA101+1/teams/#topics/solar

### Testing
- [x] i18n
- [x] RTL
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
Please check the appropriate box after all relevant reviewers have finished and given the :+1:.
- [x] Code review: @peter-fogg 
- [x] Code review: @cahrens 
- [ ] Accessibility review: @cptvitamin
- [x] Product review: @explorerleslie 
- [x] UX review: @marcotuts 
- [x] Styling / FED: @marcotuts
- [x] UI strings review: @catong

andy-armstrong commented:
jenkins run lettuce

andy-armstrong commented:
@explorerleslie @marcotuts @andywaldrop @cahrens @peter-fogg @dan-f @catong - I have team search implemented and a sandbox available for a first review:

http://andy-armstrong.sandbox.edx.org/courses/course-v1:AndyA+AA101+1/teams/#topics/solar

The CSS needs a little tweaking but other than that I believe I have all the features implemented from the wireframes.

One question: should there be a message shown when there are no results found, or is it enough to just say "Showing 0 out of 0 total"?

I'm going to start working on the tests now.

marcotuts commented:
@andy-armstrong Thanks for sharing this early view! A few notes below:

1.) Search Input styling: My suggestion would be to size the search box down to a smaller width, add placeholder text, and then some spacing between the search icon and that placeholder text. The black border-bottom was also a bit surprising to me on focus, and might be something we make a bit more subtle. It might also be good to have the border be darker on focus instead of using the blue box-shadow here.

2.) For the context label, we may want to think about this not only for search but also for empty or new topics. cc'ed @catong  - (current examples: "Showing 0 of 0 teams", "Showing 0 of 0 results", and eventually "Showing 0 of 0 posts," etc) Perhaps something like the following would work:

"There are no teams in this topic yet" (i.e. There are no {list-item-plural} in this {parent-wrapper-object}) (examples: "There are no teams in this topic, "There are no results matching this search", and eventually "There are no posts in this discussion category" etc)

andy-armstrong commented:
Thanks @marcotuts. @frrrances, I'm picking up the styling you implemented with the original static reference page, so can you comment on Marco's feedback about the border colors.

peter-fogg commented on lms/djangoapps/teams/static/teams/js/views/search_field.js:
```diff
@@ -0,0 +1,58 @@
+                },
+
+                initialize: function(options) {
+                    this.collection = options.collection;
```
This line isn't really necessary -- Backbone sets `['model', 'collection', 'el', 'id', 'attributes', 'className', 'tagName', 'events']` automatically on views when they're passed in as options.

peter-fogg commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -362,6 +365,50 @@
+                            headerModel.set(
+                                'description',
+                                interpolate(
+                                    gettext('Matching "%(searchString)s"'),
```
This string doesn't seem to be getting picked up for internationalization, though that might just be a sandbox issue.

peter-fogg commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -362,6 +365,50 @@
+                        }),
+                        view = this.createViewWithHeader({
+                            mainView: options.mainView,
+                            subject: options.subject,
```
Nit: Could just be `subject: subject`.

peter-fogg commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -362,6 +365,50 @@
+                            headerActionsView: searchFieldView
+                        }),
+                        headerModel;
+                    headerModel = view.header.model;
```
Why not condense these two lines?

peter-fogg commented on lms/djangoapps/teams/views.py:
```diff
@@ -319,7 +319,7 @@ def get(self, request):
             )
 
-        if 'text_search' in request.QUERY_PARAMS and 'order_by' in request.QUERY_PARAMS:
+        if request.QUERY_PARAMS.get('text_search', None) and request.QUERY_PARAMS.get('order_by', None):
```
Do we need to ignore query parameters with no value, like `?text_search&order_by`?

andy-armstrong replied:
Right, that's why I did this. The issue was that the server throws an error if both are provided, but Backbone will always provide both. Now it only throws an error if both are not empty.

explorerleslie commented:
@andy-armstrong looking good! A few comments: 

* I have the same comments that @marcotuts had about the styling. @frrrances let us know what you think. 
* In Notes search, we do some relative searching that Elasticsearch provides for us (matching plurals or gerunds, etc) which we don't seem to be doing here. I think it's helpful to the user to have this matching, so I would add it.

frrrances commented:
@andy-armstrong Can we check in in person when you have a minute? The style you have looks quite different than the ux template I made. Here are screenshots of what I had:

Full header with search:
<img width="1229" alt="screen shot 2015-08-26 at 11 13 16 am" src="https://cloud.githubusercontent.com/assets/4327102/9497609/cb92c8a4-4be3-11e5-977d-e07164060a6c.png">

Search with focus: 
<img width="394" alt="screen shot 2015-08-26 at 11 14 17 am" src="https://cloud.githubusercontent.com/assets/4327102/9497630/eb67067c-4be3-11e5-83ce-1a71dfbfb477.png">


Search button (icon) with hover:

<img width="291" alt="screen shot 2015-08-26 at 11 14 24 am" src="https://cloud.githubusercontent.com/assets/4327102/9497635/efa6a922-4be3-11e5-9e42-c075c942fe95.png">


Search with string entered: 

<img width="346" alt="screen shot 2015-08-26 at 11 17 22 am" src="https://cloud.githubusercontent.com/assets/4327102/9497662/1427c65a-4be4-11e5-8701-b098e9ad81a9.png">


With that said, our current focus state in the app is different so we may want to adjust that either way, but I think it is important to have the search button on the right, and indicate a hover state in case users are looking to click a button rather than hit return.

andy-armstrong commented:
@explorerleslie @frrrances @marcotuts @peter-fogg Thanks for the fast reviews. I've addressed your feedback and updated my sandbox. I've also generated i18n strings so that we can verify that those are working correctly.

@frrrances Would you be able to make the final UX tweaks on my branch? I believe that the HTML now maps correctly to your UX reference, so it should just be adjusting the focus state.

explorerleslie commented:
@andy-armstrong looking good! Two more comments: 

1. I would propose that the breadcrumb be changed to "All Topics > [Topic Name]" so that the user can easily get back to the all teams view of a topic. 
2. Similarly, @frrrances do we have a pattern for clearing search results? I seem to remember a little X in the search term box but I'm not sure.

andy-armstrong commented:
@peter-fogg @cahrens The code is now ready for review. I have implemented Jasmine tests as well as a couple of Bok Choy tests. I think I need some more Bok Choy coverage but we can discuss that tomorrow.

I also discussed with @bderusha whether we should remove the ``ENABLE_TEAMS_SEARCH`` flag as I'm not sure it makes sense to enable teams but not the search. If we don't do this, I'll need to do some further UI work to conditionalize all of the search UX based upon the setting.

catong commented:
@andy-armstrong Thanks for setting up a preview - it looks good! A couple of questions:
1) Search will only match whole words, is that right, not partial strings? 
2) What order are results shown in? (For example, in your test course, if I search "fun+test" I get 3 results)

I think the existing "Showing 0 of 0 total" message makes it clear enough that there were no matching results, but I do also like @marcotuts' suggestion for text such as "There are no results matching this search" as a nice to have. This would replace "Showing X of X total" only for the case where there are no results?

andy-armstrong commented:
@catong Great questions:
1) Unfortunately there is no stemming support in this version of search, due to limitations of the edx-search repo that we used. We will have to address this in a future release.
2) The results are shown in relevancy order, as decided by the search engine. What we would probably want is to sort equivalent results by last activity date, but that isn't easy to do. @bderusha thinks that essentially the order is random for results that are equivalent (say, those that have exactly one string match for your query). @explorerleslie, should we add a story to improve the relevancy in the future? When there are a lot of teams, this might be frustrating.

Thanks for your feedback on the message. Leslie suggests that we leave it as is for now, but if we revisit it we should use your suggested wording.

explorerleslie commented:
@andy-armstrong I'd rather see us use haystack/elasticsearch for searching, which gives us Elasticsearch's high-quality relevancy (it works really well in notes) than try to make the course search relevancy better, so I don't think it's worth adding a story in the backlog.

andy-armstrong commented:
Thanks @explorerleslie. Just to clarify, we are already using Elasticsearch through edx-search, so we are already getting the Elasticsearch relevancy. When we have a lot of teams all of which match a word, I don't think there's much that Elasticsearch can do to improve the results until we can tell it that last activity date should take part in the relevancy calculation. It doesn't seem straightforward to pass this through edx-search, so doing it when we switch to Haystack makes sense.

andy-armstrong commented:
@explorerleslie @catong I have a couple of minor questions that I keep forgetting to ask you guys:

The link at the bottom of the team page says: "Try browsing all teams or searching team descriptions." I've changed the link for 'searching team descriptions' to just scroll the user up and put their focus into the search box. Does this seem reasonable to you? Note that it only lets the user search the descriptions of teams in the current topic, not globally. I don't know if this is clear enough or whether the text should be made more explicit.

It is also a little strange that the "Browse all teams" link actually takes you to the list of all topics, from which you can browse down into teams. It doesn't feel like you are 'browsing all teams' when you click on it.

catong commented:
@andy-armstrong Has the behavior you describe for "searching team descriptions" been implemented on the current sandbox? I don't think I'm seeing that: when I'm in a topic looking at the list of teams there, and I click "searching team descriptions", I'm taken back out to the Browse tab and the topics list.

And re "browse all teams", I agree, it is confusing to be browsing topics when you've clicked "browse ... teams". 

Perhaps it's time to take a new look at the text in the footer.

andy-armstrong commented:
@catong Sorry, that wasn't clear. I haven't pushed up the new search link behavior to the sandbox yet. I have one more change to implement (Leslie's breadcrumb recommendation) and then I'll push it all up.

catong commented:
@andy-armstrong Thanks for clarifying. I edited my comment above with a suggestion for revised footer text based on the current behavior and what you've implemented for the "search" link.

explorerleslie commented:
@andy-armstrong I agree with @catong's suggestions.

cahrens commented:
I think there should be an explicit clear for search. And as it is now, if you delete the search text and move focus out (without pressing Enter), the search does clear.

cahrens commented:
There's an odd line that appears in the search box when I click in it to type.

cahrens commented on common/static/common/js/components/collections/paging_collection.js:
```diff
@@ -92,6 +98,27 @@
+                if (force || this.isStale) {
+                    this.setPage(1)
+                        .done(function() {
+                            self.isStale = false;
```
Doesn't line 92 above cover setting isStale to false?

andy-armstrong replied:
Fixed

cahrens commented on common/static/common/js/components/collections/paging_collection.js:
```diff
@@ -203,7 +230,16 @@
+            },
+
+            /**
+             * Sets the string to use for a text search. By default no search will be performed.
```
Nit: the comment "By default no search will be performed" is confusing.

andy-armstrong replied:
Fixed

cahrens commented on common/static/common/js/components/views/paging_header.js:
```diff
@@ -47,6 +47,7 @@
                 var selected = this.$('#paging-header-select option:selected');
                 this.sortOrder = selected.attr('value');
                 this.collection.setSortField(this.sortOrder);
+                this.collection.setPage(1);
```
Why do you need to set the page to 1 here? It's done in the refresh.

andy-armstrong replied:
I agree that ``refresh`` is clearer so I've switched it.

cahrens commented on common/static/common/js/components/views/search_field.js:
```diff
@@ -0,0 +1,43 @@
+                    this.$el.html(_.template(searchFieldTemplate, {
+                        type: this.type,
+                        searchString: this.collection ? this.collection.searchString : '',
+                        searchLabel: gettext('Search teams')
```
This general component in common should not have a label of "Search teams".

andy-armstrong replied:
Nice catch. I've made the label a parameter to the field now.

cahrens commented on common/static/common/js/components/views/search_field.js:
```diff
@@ -0,0 +1,43 @@
+                    event.preventDefault();
+                    if (this.collection) {
+                        this.collection.setSearchString(searchString);
+                        this.collection.setPage(1);
```
Why is it necessary to setPage to 1?

andy-armstrong replied:
I've switched this to refresh too. Note that refresh also sets the page to 1 which is useful because the search might have fewer pages than the user is currently on.

cahrens commented on common/static/common/js/spec/components/search_field_spec.js:
```diff
@@ -0,0 +1,98 @@
+
+        it('correctly displays itself', function () {
+            searchFieldView = createSearchFieldView({
+                collection: newCollection(20, 5)
```
No need to specify this specific collection.

andy-armstrong replied:
Fixed

cahrens commented on common/static/common/js/spec/components/search_field_spec.js:
```diff
@@ -0,0 +1,98 @@
+                text_search: 'foo'
+            });
+            AjaxHelpers.respondWithError(requests);
+            expect(searchFieldView.$('.search-field').val(), 'foo');
```
How can you tell that it didn't refresh? You are testing exactly the same things as on line 80.

cahrens commented on common/static/js/spec/main_requirejs.js:
```diff
@@ -155,13 +155,13 @@
         'common-requirejs/include/common/js/spec/components/paginated_view_spec.js',
         'common-requirejs/include/common/js/spec/components/paging_collection_spec.js',
         'common-requirejs/include/common/js/spec/components/paging_header_spec.js',
         'common-requirejs/include/common/js/spec/components/paging_footer_spec.js',
```
Why is view_utils_spec deleted?

andy-armstrong replied:
Bad merge. Fixed.

cahrens commented on common/test/acceptance/pages/lms/teams.py:
```diff
@@ -204,6 +207,26 @@ def click_browse_all_teams_link(self):
+        """ Search for the specified string. """
+        self.q(css='.search-field').first.fill(string)
+        self.q(css='.action-search').first.click()
+        EmptyPromise(self._showing_search_results, u"Showing search results").fulfill()
```
FYI that the new wait_for syntax is pretty nice.

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -624,8 +626,27 @@ def test_navigation_links(self):
+
+    def test_search(self):
+        """
+        Scenario:
```
Remember to update here and below.

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -624,8 +626,27 @@ def test_navigation_links(self):
+        teams = self.create_teams(self.topic, 5)
+        self.browse_teams_page.visit()
+        self.browse_teams_page.search('banana')
+        self.verify_on_page(1, teams, 'Showing 0 out of 0 total', True)
```
Can you extend these to cover clearing the search?

cahrens commented on lms/djangoapps/teams/static/teams/js/collections/team.js:
```diff
@@ -22,7 +24,12 @@
-                model: TeamModel
+                model: TeamModel,
+
+                setSearchString: function(searchString) {
```
Doesn't this method exist in the superclass?

andy-armstrong replied:
Indeed. Removed.

cahrens commented on lms/djangoapps/teams/static/teams/js/views/topic_teams.js:
```diff
@@ -22,21 +22,21 @@
+
+                            if (self.teamMemberships.canUserCreateTeam()) {
+                                var message = interpolate_text(
+                                    _.escape(gettext("Try {browse_span_start}browsing all teams{span_end} or {search_span_start}searching team descriptions{span_end}. If you still can't find a team to join, {create_span_start}create a new team in this topic{span_end}.")),
```
This string (which I know you did not change) will be very hard for translators to deal with.

cahrens commented:
Clicking "searching team descriptions" does not work.

We are also going to have to figure out how to internationalize that string in a reasonable way....

cahrens commented on lms/djangoapps/teams/static/teams/js/views/topic_teams.js:
```diff
@@ -47,8 +47,10 @@
                     event.preventDefault();
-                    // TODO! Will navigate to correct place once required functionality is available
-                    Backbone.history.navigate('browse', {trigger: true});
+                    $('.page-header-search .search-field').focus();
```
This didn't work for me when I clicked on the text at the end of the list of teams.

andy-armstrong replied:
This wasn't on the sandbox at the time. It is there now.

cahrens commented on lms/djangoapps/teams/static/teams/js/spec/views/teams_tab_spec.js:
```diff
@@ -137,5 +137,92 @@ define([
+                expect(teamsTabView.$('.page-description').text()).toBe('Showing results for "foo"');
+            });
+
+            it('can clear a search', function () {
```
Personally, I would just combine this with the previous test point so you can be sure that the setup was correct (before the clear).

andy-armstrong commented:
@explorerleslie @catong @cahrens @peter-fogg I've completed the code changes and updated the sandbox. Now the search breadcrumbs behave as recommended by @explorerleslie, and I've updated the browse and search links. 

I have a number of new test failures that I'll investigate in the morning, but for now I think the functionality is all set. @frrrances, could you address the styling for the search field focus (the extra black line that folks don't like).

explorerleslie commented:
:+1:

andy-armstrong commented:
@cahrens
> I think there should be an explicit clear for search. And as it is now, if you delete the search text
> and move focus out (without pressing Enter), the search does clear.

At @explorerleslie's suggestion, I've changed search to have its own navigation state, so that the breadcrumbs have a link back to the full list. I've also implemented searching on focus out so you can clear it that way too. Let me know if that satisfies your concerns. (this is live on the sandbox now)

andy-armstrong commented:
@cahrens @explorerleslie @marcotuts @frrrances I've updated my sandbox to remove the black line in the search field.

peter-fogg commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -1,9 +1,10 @@
-;(function (define) {
+;(function (define, interpolate) {
```
Doesn't `interpolate` come along with `gettext`?

andy-armstrong replied:
It does, but JSHint doesn't understand this. It means that my code gets error underlines on every usage. I was experimenting with whether this was cleaner, because then I only have one error at the bottom of the page.

andy-armstrong replied:
I now understand your point better... if ``interpolate`` was brought in on demand, then it wouldn't be defined at this point. I'm pretty sure that it is defined in the environment and that ``gettext`` just provides a shim, but it does make it look confusing. I wish there were a cleaner way to bring it in too. Maybe we should provide it as a shim too.

cahrens commented on common/static/common/js/spec/components/search_field_spec.js:
```diff
@@ -0,0 +1,96 @@
+                start: 0,
+                results: []
+            });
+            expect(searchFieldView.$('.search-field').val(), 'foo');
```
@andy-armstrong I'm not sure you saw my note below on line 93. I don't see a difference between this test point and the one below, with regard to whether or not something refreshed.

andy-armstrong replied:
Yep, I did see it. I'm still thinking about how I can test it properly.

andy-armstrong replied:
I've decided to move the test out of here and into the Teams Tab spec. The bare class itself can't really do much about handling errors, so it makes more sense to have the error handling tested within a page.

andy-armstrong commented:
@cahrens I've refactored away the confusing team params concept, and instead pass around the server's context between views. I think this is much clearer so I hope this is more in line with what you were thinking. It turned into quite a large change as all the tests had to be updated to pass the context correctly, so I moved that logic into team_spec_helpers so that it could be easily shared.

https://github.com/edx/edx-platform/commit/d114fca53cf5e278d494a173e4d0fa02d756ca9e

I still have to deal with the Bok Choy tests so that will have to happen on Monday.

catong commented:
@andy-armstrong The updated search field and X look great. I noticed a couple of things:
- The search icon isn't aligned vertically with the search field, and looks odd. 
![searchfieldalign](https://cloud.githubusercontent.com/assets/9041399/9585721/97f03cb4-4fe6-11e5-8157-3b51abf1b3b9.png)

- I also wonder about the experience when I get no results on a search. The empty search results page has no indication of what topic I was in when I searched previously, so while I can follow the suggestion to "search team descriptions" in this topic, I can't see what topic I'm currently searching in. The focus goes back up to the search field, which is great, but I wonder if we could also refresh the page so that we're back on the initial list of teams for a specific topic but with the focus in the search field?

andy-armstrong commented:
@catong Thanks for the feedback:
1. @frrrances just gave me improved CSS rules for the check box, so it should be better shortly.
2. You should see the topic you searched from in the breadcrumbs. Do you think it needs further calling out?

catong commented:
Thanks @andy-armstrong, I wasn't very clear in my comment. When I do a search and get no results, I guess I'm one level down from "All Topics", there is no topic shown in the breadcrumbs. See screenshots. Do you see the same?
![searchnoresults](https://cloud.githubusercontent.com/assets/9041399/9586151/0e293fc8-4fe9-11e5-9048-4f3b80b033e4.png)

catong commented:
@andy-armstrong We just confirmed that what I saw (lack of topic in breadcrumbs) was some aberration and not reproducible, and the revised styling got rid of the alignment issue, so both my questions are addressed. Thanks!

andy-armstrong commented:
jenkins run lettuce

andy-armstrong commented:
@cahrens @peter-fogg I believe I have all the tests passing, so please go ahead and re-review.

peter-fogg commented on common/static/common/js/components/collections/paging_collection.js:
```diff
@@ -92,6 +102,26 @@
+             * @param force If true, it will always refresh.
+             * @returns {promise} Returns a promise representing the refresh.
+             */
+            refresh: function(force) {
```
As far as I can tell we're not using the `force` option anywhere. Do we need to keep it?

peter-fogg commented on common/static/common/js/components/views/search_field.js:
```diff
@@ -0,0 +1,50 @@
+                },
+
+                initialize: function(options) {
+                    Backbone.View.prototype.initialize.call(this, options);
```
`Backbone.View.prototype.initialize` is just `function () {}`, so there's no need to make this call.

andy-armstrong replied:
I like the convention of ``initialize`` methods always calling super, so that they don't accidentally skip over needed initialization. I do this in other languages and figured I'd do it here, but you are right it isn't needed. What happens if a future version of Backbone adds some initialization?

I can remove it if you really don't like it...

cahrens replied:
I don't feel strongly either way about this, but we should be consistent. Always calling super seems to be the easiest mental model.

peter-fogg replied:
I just don't see the need for the extra line noise and method call. I can't imagine Backbone would change `initialize` in a future version; that's what `constructor` is for.

peter-fogg commented on common/static/common/js/components/views/search_field.js:
```diff
@@ -0,0 +1,50 @@
+                    Backbone.View.prototype.initialize.call(this, options);
+                    this.type = options.type;
+                    this.label = options.label;
+                    _.bindAll(this, 'performSearch');
```
Backbone should automatically bind `performSearch` if it's being called as an event handler.

andy-armstrong replied:
I'd been wondering about this. A lot of the teams code has this line, and once when I tried removing it it broke some tests. I'll remove them as I come across them and see if I can reproduce the issue.

peter-fogg replied:
I tested this out in #9559, and all tests are passing.

peter-fogg commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -503,10 +503,18 @@ def teams_with_default_sort_order(self, teams):
+            'Showing results for "{search_query}"'.format(search_query=search_query)
+        )
+
+    def verify_on_page(self, teams_page, page_num, total_teams, pagination_header_text, footer_visible):
```
Update the documentation for the `teams_page` argument. Also, is there any sort of type constraint for `teams_page`? It would be nice if there were some mixin which implemented the necessary methods. If each class requires different method implementations then I suppose there could be a common abstract base class, but maybe that's overengineering it somewhat...

andy-armstrong replied:
It has to be a subclass of ``BaseTeamsPage`` so I've specified that in the docstring. Let me know if you'd rather that I break it into a mixin.

peter-fogg commented on lms/djangoapps/teams/static/teams/js/spec_helpers/team_spec_helpers.js:
```diff
@@ -105,7 +108,7 @@ define([
                 team_memberships_data: createMockTeamMembershipsData(1, 5)
             },
-            options
+            options || {}
```
`_.extend` treats non-object parameters as empty objects, so this change isn't necessary.

andy-armstrong replied:
Thanks! The documentation doesn't make this clear...

http://underscorejs.org/#extend

peter-fogg commented on lms/djangoapps/teams/static/teams/js/spec_helpers/team_spec_helpers.js:
```diff
@@ -236,14 +239,49 @@ define([
+                name: 'Test Topic 1',
+                description: 'Test description 1'
+            },
+            options || {}
```
Same note about `_.extend` here.

peter-fogg commented on lms/djangoapps/teams/static/teams/js/spec_helpers/team_spec_helpers.js:
```diff
@@ -236,14 +239,49 @@ define([
+    };
+
+    var createMockContext = function(options) {
+        return _.extend(_.extend({}, testContext), options);
```
This can be folded up into `_.extend({}, testContext, options)`.

andy-armstrong replied:
Nice! Another thing I didn't notice in the documentation...

peter-fogg commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -21,24 +21,21 @@
-                    this.courseID = options.teamParams.courseID;
-                    this.topicID = options.teamParams.topicID;
+                    this.context = options.context;
+                    this.topic = options.topic;
```
What dictates which information should go into the `context

andy-armstrong replied:
The context is what is passed down from the server to the factory class. I'm trying to treat is as immutable so any dynamic values get passed in separately.

peter-fogg commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -382,52 +374,55 @@
                         });
                     }
-                    if (options.parentTeam) {
+                    if (topic && team) {
```
Nit: push this down into the `if (topic)` block, since there we already know that the first argument to the `&&` is truthy.

peter-fogg commented:
@catong @explorerleslie @andy-armstrong The footer on the teams-for-topic page has the text "... or search team descriptions in this topic." Should that be changed to reflect that we search across name, country, and language as well?

peter-fogg commented:
@andy-armstrong I'm finished with my review -- looks great.

andy-armstrong commented:
@peter-fogg I've addressed your feedback, and rebased to fix a merge conflict with the ``is_active`` changes.

@cahrens This is ready for your review. I updated the sandbox and noticed that I've somehow broken the styling for the clear search button when I squashed commits. I'll resurrect that shortly, but everything else should be all set.

catong commented:
@peter-fogg Thanks for bringing that up about the text in the footer. I have been documenting the search as only working on team descriptions because of that! :-) I would definitely be in favor of updating the string to be more accurate. Should we just say "search teams in this topic"?
@andy-armstrong @explorerleslie ?

cahrens commented on common/static/common/js/components/views/search_field.js:
```diff
@@ -0,0 +1,49 @@
+                },
+
+                performSearch: function(event) {
+                    var searchField = this.$('.search-field'),
```
Declare dependency on jQuery (for some reason, we have left that out very often in the teams JS files... I'm adding a bunch in with my PR).

cahrens commented on common/static/common/js/components/views/paging_header.js:
```diff
@@ -47,6 +47,7 @@
                 var selected = this.$('#paging-header-select option:selected');
                 this.sortOrder = selected.attr('value');
                 this.collection.setSortField(this.sortOrder);
+                this.collection.refresh();
```
return this.collection.refresh();

Just in case anyone wants the promise, and to be consistent with below.

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -503,15 +503,24 @@ def teams_with_default_sort_order(self, teams):
 
-    def verify_on_page(self, page_num, total_teams, pagination_header_text, footer_visible):
+    def verify_search_header(self, search_results_page, search_query):
+        """Verify that the page header correctly reflects the current topic's name and description."""
```
Update doc string.

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -503,15 +503,24 @@ def teams_with_default_sort_order(self, teams):
 
         Arguments:
-            page_num (int): The one-indexed page we expect to be on
+            page (BaseTeamsPage): The teams page object that should be the current page.
```
Should be teams_page.

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -670,36 +679,48 @@ def test_teams_page_input(self):
+        When I should navigate to that link
+        Then I should see the topic browse page
         """
+        self.create_teams(self.topic, self.TEAMS_PAGE_SIZE + 10)
```
Did this "work" before just because of ordering?

andy-armstrong replied:
It doesn't strictly need teams because it clicks on the link at the bottom of the page which is there anyway. I added it when I was trying to get the search link to correctly set the focus, but I then decided it was better to test that in Jasmine.

I'll remove this if it works without the change.

cahrens commented on lms/djangoapps/teams/static/teams/js/spec/views/edit_team_spec.js:
```diff
@@ -13,21 +13,21 @@ define([
-                description: "TeamDescription",
-                country: "US",
-                language: "en",
+                name: 'TeamName',
```
Nit: personally I don't care about mixing single and double quotes, and switching existing quotes makes reviews harder. My preference would be not to change them unnecessarily (not that you need to change what you've already done in this PR... a suggestion for going forward).

andy-armstrong replied:
I'll try to restrain myself in future. :-)

cahrens commented on lms/djangoapps/teams/static/teams/js/spec/views/team_profile_spec.js:
```diff
@@ -11,7 +11,7 @@ define([
                 {
                     'user': {
-                        'username': 'bilbo',
+                        'username': TeamSpecHelpers.testUser,
```
Nit: this test is now a little less fun to read. :)

andy-armstrong replied:
I agree, but it makes them easier to write. :-)

cahrens commented on lms/djangoapps/teams/static/teams/js/spec/views/teams_tab_spec.js:
```diff
@@ -137,5 +116,103 @@ define([
+
+                // Clear the search and submit it again
+                teamsTabView.$('.search-field').val('');
+                teamsTabView.$('.action-search').click();
```
Will also need a test for the "clear" button. You could refactor these 3 similar test cases (around clearing) to share code, if so desired...

andy-armstrong replied:
I already have Jasmine tests for the clear button in the search field spec, so I figured that I didn't have to re-test all the combinations here. I'll look into creating a shared method and then maybe having all three tests won't seem duplicative.

cahrens commented on lms/djangoapps/teams/static/teams/js/spec/views/topic_teams_spec.js:
```diff
@@ -65,17 +62,16 @@ define([
-            // TODO! Should be updated once team description search feature is available
-            expect(Backbone.history.navigate.calls[0].args).toContain('browse');
+            spyOn($.fn, 'focus').andCallThrough();
+            teamsView.$('.search-teams').click();
```
Should it do anything with text that is already in the search box? The caret position seems odd now.

andy-armstrong replied:
Good point. It should probably select the text to make it easier to replace.

cahrens commented on lms/djangoapps/teams/static/teams/js/spec/views/topic_teams_spec.js:
```diff
@@ -118,13 +116,13 @@ define([
             teamMemberships.teamEvents.trigger('teams:update', { action: 'create' });
             teamsView.render();
-            AjaxHelpers.expectJsonRequestURL(
+            AjaxHelpers.expectRequestURL(
```
Should we go ahead and delete this commented out test point?

cahrens commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -267,54 +254,70 @@
+                            description: options.description,
+                            breadcrumbs: options.breadcrumbs
+                        });
+                    this.listenTo(collection, 'sync', function() {
```
When is this hit?

andy-armstrong replied:
I've added a comment explaining this.

cahrens replied:
But when will the collection by synced and yet we still want to maintain the search? Can you explain the reproduction steps that would result in a sync?

cahrens commented on lms/djangoapps/teams/static/teams/js/views/topic_teams.js:
```diff
@@ -25,21 +26,21 @@
-                    });
+                            if (self.teamMemberships.canUserCreateTeam()) {
+                                var message = interpolate_text(
+                                    _.escape(gettext("{browse_span_start}Browse teams in other topics{span_end} or {search_span_start}search team descriptions{span_end} in this topic. If you still can't find a team to join, {create_span_start}create a new team in this topic{span_end}.")),
```
This is pretty complex for translators. Anything we can do to improve it? At the very least, could use a translator comment.

cahrens commented:
Wow, this was an exhausting amount of code to review-- great work! Just a few suggestions/questions, but otherwise :+1: It would be great to get this in soon since @peter-fogg  and I will both need to change our PRs accordingly.

Of course, I'm still seeing the issue with the clear and search buttons both appearing, but I figured you just haven't gotten the CSS cleaned up.

cahrens commented:
And the bok choy test situation is very confusing. Only 2 tests ran?

andy-armstrong commented:
@frrrances How would you like to handle the UX review for your changes? I'd like to merge this tomorrow morning if possible.

frrrances commented on lms/static/sass/views/_teams.scss:
```diff
@@ -83,41 +84,56 @@
-      padding: ($baseline/4) ($baseline/2);
-
+      padding: ($baseline/5) ($baseline/2);
+      vertsical-align: middle;
```
Looks like a typo crept in here.. should be "vertical-align:"

frrrances commented:
@marcotuts Any chance you could take a quick spin through my sass changes so I'm not patting myself on my own back?

peter-fogg commented:
@andy-armstrong :+1:

andy-armstrong commented:
@marcotuts I've resurrected my sandbox and verified that the styling issue with the clear search button has been resolved. It should be ready for your review now. Thanks!

http://andy-armstrong.sandbox.edx.org/courses/course-v1:AndyA+AA101+1/teams/#topics/solar

marcotuts commented on lms/static/sass/views/_teams.scss:
```diff
@@ -252,6 +268,11 @@
         @include margin-right ($baseline*.75);
         color: $gray-d1;
 
+        abbr {
```
I couldn't find this being used in the templates for this PR, where is this abbr used? cc'ed @frrrances

marcotuts commented on lms/static/sass/views/_teams.scss:
```diff
@@ -49,16 +49,17 @@
-        @include text-align(right);
-      }
+    .page-header-secondary {
+      display: inline-block;
```
NIT: property ordering here, include should move to the top.

marcotuts commented on lms/static/sass/views/_teams.scss:
```diff
@@ -83,41 +84,56 @@
   .page-header-search {
 
+    .wrapper-search-input {
+      position: relative;
```
NIT: property odering here: display --> position --> vertical align

marcotuts commented on lms/static/sass/views/_teams.scss:
```diff
@@ -83,41 +84,56 @@
       @extend %button-reset;
-      padding: ($baseline/4) ($baseline/2);
-
+      padding: ($baseline/5) ($baseline/2);
```
NIT: property order: extend --> background-color --> padding --> text-shadow

andy-armstrong commented:
@frrrances @marcotuts I just discovered that the clear button doesn't work with right-to-left:

http://andy-armstrong.sandbox.edx.org/courses/course-v1:AndyA+AA101+1/teams/?preview-lang=rtl#topics/solar

![image](https://cloud.githubusercontent.com/assets/5985072/9635855/6d118e68-5166-11e5-9121-5cb380a8d29d.png)

marcotuts commented:
A couple of small nits on property ordering, but nothing else from me. (Also want to make sure we fix the vertical-align typoe frances mentioned.) :+1:

andy-armstrong commented:
Thanks @marcotuts. I've squashed commits and included the typo fix.

marcotuts commented on lms/static/sass/views/_teams.scss:
```diff
@@ -83,41 +84,56 @@
+      @extend %button-reset;
+      position: absolute;
+      top: 0;
+      right: 0;
```
Great catch on the RTL styling @andy-armstrong. This line should be @include right(0);

andy-armstrong replied:
Thanks @marcotuts. That means that the line should move to the top as includes go first, yes?

marcotuts commented on lms/static/sass/views/_teams.scss:
```diff
@@ -83,41 +84,56 @@
+      position: absolute;
+      top: 0;
+      right: 0;
+      margin: 0 ($baseline/4) 0 0;
```
Similarly, this line should be @include margin(0, ($baseline/4), 0, 0);

andy-armstrong commented:
@marcotuts I've pushed up changes to your feedback and I'm rebuilding the sandbox.

andy-armstrong commented:
@catong FYI, I changed the string to use your suggestion of "search teams in this topic".

catong commented:
@andy-armstrong Thanks! I added a checkbox for UI strings review to the cover comment and marked it done. :+1: for UI strings.
